daft/execution/execution_step.py

```diff
@@ -66,85 +78,68 @@ def add_instruction(
         self,
         instruction: Instruction,
         resource_request: ResourceRequest | None,
-    ) -> OpenExecutionQueue[PartitionT]:
+    ) -> ExecutionStepBuilder[PartitionT]:
         """Append an instruction to this ExecutionStep's pipeline."""
         self.instructions.append(instruction)
         self.resource_request = ResourceRequest.max_resources([self.resource_request, resource_request])
         return self
 
-    def as_materialization_request(self) -> MaterializationRequest[PartitionT]:
-        """Create an MaterializationRequest from this ExecutionStep.
+    def build_materialization_request_single(self) -> ExecutionStepSingle[PartitionT]:
+        """Create an ExecutionStepSingle from this ExecutionStepBuilder.
 
         Returns a "frozen" version of this ExecutionStep that cannot have instructions added.
         """
-        return MaterializationRequest[PartitionT](
+        return ExecutionStepSingle[PartitionT](
             inputs=self.inputs,
             instructions=self.instructions,
             num_results=1,
             resource_request=self.resource_request,
         )
 
-    def as_materialization_request_multi(self, num_results: int) -> MaterializationRequestMulti[PartitionT]:
-        """Create an MaterializationRequestMulti from this ExecutionStep.
+    def build_materialization_request_multi(self, num_results: int) -> ExecutionStepMulti[PartitionT]:
+        """Create an ExecutionStepMulti from this ExecutionStepBuilder.
 
         Same as as_materization_request, except the output of this ExecutionStep is a list of partitions.
         This is intended for execution steps that do a fanout.
         """
-        return MaterializationRequestMulti[PartitionT](
+        return ExecutionStepMulti[PartitionT](
             inputs=self.inputs,
             instructions=self.instructions,
             num_results=num_results,
             resource_request=self.resource_request,
         )
 
-
-@dataclass
-class MaterializationRequestBase(ExecutionStep[PartitionT]):
-    """Common helpers for MaterializationRequest and MaterializationRequestMulti.
-    See those classes for more details.
-
-    num_results: The number of partitions that will be returned.
-    _id: A unique identifier for this ExecutionStep.
-    """
-
-    num_results: int
-    _id: int = field(default_factory=lambda: next(ID_GEN))
-
-    def id(self) -> str:
-        return f"{self.__class__.__name__}_{self._id}"
-
     def __str__(self) -> str:
-
         return (
-            f"{self.id()}\n"
+            f"ExecutionStepBuilder\n"
             f"  Inputs: {self.inputs}\n"
             f"  Resource Request: {self.resource_request}\n"
             f"  Instructions: {[i.__class__.__name__ for i in self.instructions]}"
         )
 
 
 @dataclass
-class MaterializationRequest(MaterializationRequestBase[PartitionT]):
+class ExecutionStepSingle(ExecutionStep[PartitionT]):
```

Comment:
Good point!

Perhaps `SingleOutputExecutionStep` and `MultiOutputExecutionStep`?

Separate question - why do we need subclasses to differentiate these two since we already have `.num_results=1/many` to differentiate the two? Could every `ExecutionStep` not just return a list of results (potentially with length 1)?